hnix-store-json/src/System/Nix/JSON.hs

```diff
@@ -53,70 +64,160 @@ instance FromJSON StorePathHashPart where
     . System.Nix.Base.decodeWith NixBase32
     )
 
+-- | TODO: hacky, we need to stop assuming StoreDir for
+-- StorePath to and from JSON
 instance ToJSON StorePath where
   toJSON =
     toJSON
-    -- TODO: hacky, we need to stop requiring StoreDir for
-    -- StorePath rendering and have a distinct
-    -- types for rooted|unrooted paths
-    . Data.Text.drop 1
-    . System.Nix.StorePath.storePathToText (StoreDir mempty)
+    . System.Nix.StorePath.storePathToText def
 
   toEncoding =
     toEncoding
-    . Data.Text.drop 1
-    . System.Nix.StorePath.storePathToText (StoreDir mempty)
+    . System.Nix.StorePath.storePathToText def
 
 instance FromJSON StorePath where
   parseJSON =
     withText "StorePath"
     ( either
         (fail . show)
         pure
-    . System.Nix.StorePath.parsePathFromText (StoreDir mempty)
-    . Data.Text.cons '/'
+    . System.Nix.StorePath.parsePathFromText def
     )
 
-instance ToJSON (DerivationOutput OutputName) where
+instance ToJSONKey StorePath where
+  toJSONKey = Data.Aeson.Types.toJSONKeyText $ System.Nix.StorePath.storePathToText def
+
+instance FromJSONKey StorePath where
+  fromJSONKey = FromJSONKeyTextParser $ either (fail . show) pure . System.Nix.StorePath.parsePathFromText def
```

Comment:
the `def` here is evil (assumes `/nix/store`)

As of https://github.com/NixOS/nix/pull/13980 and https://github.com/NixOS/nix/issues/13570, should not do that instead, just skip the store dir.